### PR TITLE
Fix TBB Threading issue

### LINKED ISSUE #26703: Error when running systems tests on Windows Debug build
### Expected behavior
Running system tests with a debug build on Windows should work,

### Actual behavior
When you run a set of system tests, the tests will run correctly and the outputs will tell you that `CompareWorkspaces` was successful for each test. However, at the end of the session, an error will occur and the output will say that all tests have failed:
```
0% tests passed, 29 tests failed out of 29 (494 skipped)
All tests passed? False
################################################################################
Assertion is_set(s) failed on line 215 of file ../../src/tbb/governor.cpp
Detailed description: Attempt to terminate non-local scheduler instance
```

### Steps to reproduce the behavior
- Build in Debug mode on Windows
- Open `command-prompt.bat` in the build folder
- Run a set of system tests e.g. `systemtest.bat -C Debug -R ISISIndirectInelastic.IRISConvFit`

It seems to be an issue with TBB.
Googling the error lead me to: https://github.com/ospray/ospray/issues/282 where someone had the exact same error. Unhelpfully, at time of writing, no one had responded.

### Platforms affected
Windows

### PR DESCRIPTION
**Pull request by DavidFair**
The text below / in the commit is more for anyone else searching this error message, since apart from us and the ospray project, nobody else has seen this assert.

-----

Fixes the following assertion thrown by TBB:

	Attempt to terminate non-local scheduler instance

This took a significant amount of time to track down due to a number of
factors. If your looking for a quick fix; set your
tbb::task_scheduler_init to use thread local storage.

This is caused by using Python multi-processing (in our case for the
system tests). The process will spawn correctly (Windows), however
Python will helpfully create seperate thread(s). I assume to observe. In
this case the task_scheduler_init handle would then belong to one of
these threads, as the act of importing our Python module would init the
framework and TBB implicitly.

When the tests run and TBBs scheduling is required it will "sign_on" in
tbb's governor.cpp to handle scheduling. This works because the
scheduler_init handle
is process wide. However internally they then use thread local storage
(TLS) for tracking the scheduler.

As the main process goes to destruct it assumes that we must have
a scheduler instance, since at process level we have done the init.

But we cannot see any schedulers since they belong to a different
thread. We have been getting away with this as the TLS returns NULL
which effectively nops the destruction. But the assert correctly points
out were getting this wrong.

Marking each handle to the scheduler as thread local effectively ties
the destruction of the thread to its init status. For most threads this
effectively means we create a unique_ptr which isn't used. But for our
main threads and weird edge case worker threads it keeps the thread
state consistent.

We've used a unique_ptr since we have an API which allows us to change
the number of threads TBB can spin up. However, anyone else facing this
issue would probably be fine with a thread local object allocated on the
stack.

-----

**To test:**
- Ensure `Testing/SystemTests/lib/systemtests` is in your PYTHONPATH
- Run the following minimum reproducer and check it does not throw an assert
```python
from multiprocessing import Manager

def main():
    from systemtesting import TestSuite

    # reproduced from tmgr.loadTestsFromDir then loadTestsFromModule
    # i.e. loading a TestSuite in systemtesting.py
    with Manager() as m:
        tests_dict = m.dict()

        # Is this TestSuite in a different thread? Either way BOOM!
        tests_dict['0'] = TestSuite(None, None, None, None)
        tests_dict['0'] = None

if __name__ == "__main__":
    main()
```
- Run the following system test `.\systemtest.bat -C Debug -R OSIRISConv`

Fixes #26703 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DanNixon** [comment]
Perhaps it would be worth looking to replace this code as it uses deprecated TBB features: https://software.intel.com/en-us/node/506296.

**Comment by martyngigg** [comment]
@DanNixon David has updated the TBB version in #28372 but I guess we might not have access to the new API on all platforms? Shall we merge this so it fixes the bug and look at if migration to the new api is possible for all platforms wuth #28732?

**Comment by DavidFair** [comment]
The caveat being we lose control of the number of threads TBB uses, since it would determine it automatically

I guess the question should be, is there any actual uses cases for not letting TBB figure it out and using the value from the config service?

**Comment by DanNixon** [comment]
> I guess the question should be, is there any actual uses cases for not letting TBB figure it out and using the value from the config service?

Yes, 100% yes. This becomes a serious problem when you start combining several parts of Mantid (.e.g live streaming, reduction and instrument view).

The replacement API does allow you to control how many threads are used so I'm not sure I understand the caveat you talk about.

**Comment by DavidFair** [comment]
Edit I can see this file [global_control](https://software.intel.com/en-us/node/589744) is the direct replacement

We can drop thread local and the unique_ptr dance by doing this, so I'm happy to re-order the 2020 update first then the PR?

**Comment by martyngigg** [comment]
It looks like we'll need to update the [RHEL version](https://copr.fedorainfracloud.org/coprs/mantid/mantid/build/446358/). I think it's too old to contain `global_control`. If that's too difficult we'll have to `#ifdef` the code based on the version.

**Comment by DavidFair** [comment]
> 
> 
> It looks like we'll need to update the [RHEL version](https://copr.fedorainfracloud.org/coprs/mantid/mantid/build/446358/). I think it's too old to contain `global_control`. If that's too difficult we'll have to `#ifdef` the code based on the version.

Probably worth updating it in the other PR, so all the updates are together

**Comment by martyngigg** [comment]
> > It looks like we'll need to update the [RHEL version](https://copr.fedorainfracloud.org/coprs/mantid/mantid/build/446358/). I think it's too old to contain `global_control`. If that's too difficult we'll have to `#ifdef` the code based on the version.
> 
> Probably worth updating it in the other PR, so all the updates are together

:+1: We'll just need to update the `mantid-developer.spec` to set a minimum version and then we should set that minimum version in CMake too so it bails out early if the version is too old. I've kicked off a [test build](https://copr.fedorainfracloud.org/coprs/martyngigg/sandbox/build/1304766/) of the latest version for EL7. Hopefully it will just work.

**Comment by DavidFair** [comment]
I think were going do have to `#ifdef` this since Ubuntu have TBB 2017, whereas `global_control` was introduced in TBB 2018

File: Framework/API/src/FrameworkManager.cpp [hunk 1]
```diff
@@ -22,10 +22,22 @@
...
 
 #include <clocale>
 #include <cstdarg>
+#include <memory>
+
+// The below can be replaced with global_control when all platforms
+// have 2019U4
+#include "tbb/tbb_stddef.h"
+
+#if TBB_INTERFACE_VERSION_MAJOR < 11
```
Comment (martyngigg) [inline Framework/API/src/FrameworkManager.cpp]: Can we use `__has_include(<tbb/global_control.h>)` now?

Comment (DavidFair) [reply Framework/API/src/FrameworkManager.cpp]: I had a look at this, however Ubuntu has `tbb/global_control.h`but also requires us to do something like `#define PREVIEW_GLOBAL_CONTROL`

By marking it as TBB version 11+ we avoid having to mess with pre-processor defines too

Comment (martyngigg) [reply Framework/API/src/FrameworkManager.cpp]: Sounds fair. Thanks.

**Comment by martyngigg** [comment]
The RHEL failure is fixed by #28375

**Review by martyngigg — APPROVED** [review]
This looks good to me now. The tests pass without an error being raised. Nice work!

**Pull request merged**